File: fcm_digital_twin/fcm_digital_twin/scripts/alara_speed_reflex.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid
from nav2_msgs.msg import SpeedLimit
from rclpy.qos import QoSProfile, QoSDurabilityPolicy
import tf2_ros

class AlaraSpeedReflex(Node):
    def __init__(self):
        super().__init__('alara_speed_reflex')

        self.rad_threshold = 20.0        
        
        # Задаем скорости в ПРОЦЕНТАХ от максимума в navigation.yaml
        self.normal_speed_pct = 35.0  # Спокойная езда (35% от 1.2 м/с = ~0.4 м/с)
        self.boost_speed_pct = 100.0  # ГАЗ В ПОЛ! (100% от 1.2 м/с)

        self.is_boosted = False
        self.latest_map = None

        map_qos = QoSProfile(depth=1, durability=QoSDurabilityPolicy.TRANSIENT_LOCAL)
        self.map_sub = self.create_subscription(OccupancyGrid, '/radiation_map', self.map_callback, map_qos)

        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer, self)

        # ОФИЦИАЛЬНЫЙ ТОПИК NAV2 ДЛЯ ОГРАНИЧЕНИЯ СКОРОСТИ
        self.limit_pub = self.create_publisher(SpeedLimit, '/speed_limit', 10)
        
        self.timer_check = self.create_timer(0.1, self.check_reflex)
        self.get_logger().info("ALARA Native Speed Limit Reflex started!")

    def map_callback(self, msg):
        self.latest_map = msg

    def publish_speed_limit(self, pct):
        msg = SpeedLimit()
        msg.percentage = True     # Говорим Nav2, что это проценты, а не м/с
        msg.speed_limit = pct
        self.limit_pub.publish(msg)

    def check_reflex(self):
        if self.latest_map is None: return

        try:
            trans = self.tf_buffer.lookup_transform('map', 'base_link', rclpy.time.Time())
            rx = trans.transform.translation.x
            ry = trans.transform.translation.y
        except Exception: return

        res = self.latest_map.info.resolution
        ox = self.latest_map.info.origin.position.x
        oy = self.latest_map.info.origin.position.y
        width = self.latest_map.info.width

        rad_x = int((rx - ox) / res)
        rad_y = int((ry - oy) / res)

        if rad_x < 0 or rad_y < 0 or rad_x >= width or rad_y >= self.latest_map.info.height: return
        rad_val = self.latest_map.data[rad_y * width + rad_x]

        # Логика переключения
        if rad_val >= self.rad_threshold and not self.is_boosted:
            self.get_logger().warn(f"RADIATION ({rad_val}%)! BOOSTING TO {self.boost_speed_pct}% SPEED!")
            self.is_boosted = True
            
        elif rad_val < self.rad_threshold and self.is_boosted:
            self.get_logger().info(f"Radiation cleared. Restoring limit to {self.normal_speed_pct}%.")
            self.is_boosted = False

        # Постоянно шлем лимит, чтобы Nav2 держал его (требование Nav2)
        if self.is_boosted:
            self.publish_speed_limit(self.boost_speed_pct)
        else:
            self.publish_speed_limit(self.normal_speed_pct)

def main(args=None):
    rclpy.init(args=args)
    node = AlaraSpeedReflex()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: fcm_costmap_plugins/src/alara_speed_reflex.cpp
#include "rclcpp/rclcpp.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "nav2_msgs/msg/speed_limit.hpp"
#include "tf2_ros/buffer.h"
#include "tf2_ros/transform_listener.h"
#include "geometry_msgs/msg/transform_stamped.hpp"

using std::placeholders::_1;
using namespace std::chrono_literals;

class AlaraSpeedReflex : public rclcpp::Node
{
public:
  AlaraSpeedReflex()
  : Node("alara_speed_reflex_cpp"),
    rad_threshold_(30.0),
    normal_speed_pct_(35.0),
    boost_speed_pct_(100.0),
    is_boosted_(false),
    init_done_(false)
  {
    // Transient Local QoS
    auto qos = rclcpp::QoS(rclcpp::KeepLast(1)).transient_local();

    map_sub_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
      "/radiation_map", qos, std::bind(&AlaraSpeedReflex::map_callback, this, _1));

    limit_pub_ = this->create_publisher<nav2_msgs::msg::SpeedLimit>("/speed_limit", 10);

    tf_buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
    tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

    timer_check_ = this->create_wall_timer(
      500ms, std::bind(&AlaraSpeedReflex::check_reflex, this));

    timer_init_ = this->create_wall_timer(
      1000ms, std::bind(&AlaraSpeedReflex::init_speed, this));

    RCLCPP_INFO(this->get_logger(), "ALARA Native C++ Reflex started! (Zero CPU load)");
  }

private:
  void init_speed()
  {
    if (!init_done_) {
      publish_speed_limit(normal_speed_pct_);
      init_done_ = true;
    }
  }

  void map_callback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg)
  {
    latest_map_ = msg;
  }

  void publish_speed_limit(double pct)
  {
    nav2_msgs::msg::SpeedLimit msg;
    msg.percentage = true;
    msg.speed_limit = pct;
    limit_pub_->publish(msg);
  }

  void check_reflex()
  {
    if (!latest_map_) return;

    geometry_msgs::msg::TransformStamped trans;
    try {
      trans = tf_buffer_->lookupTransform("map", "base_link", tf2::TimePointZero);
    } catch (const tf2::TransformException & ex) {
      return; 
    }

    double rx = trans.transform.translation.x;
    double ry = trans.transform.translation.y;

    double res = latest_map_->info.resolution;
    double ox = latest_map_->info.origin.position.x;
    double oy = latest_map_->info.origin.position.y;
    int width = latest_map_->info.width;
    int height = latest_map_->info.height;

    int rad_x = static_cast<int>((rx - ox) / res);
    int rad_y = static_cast<int>((ry - oy) / res);

    if (rad_x < 0 || rad_y < 0 || rad_x >= width || rad_y >= height) return;

    int8_t rad_val = latest_map_->data[rad_y * width + rad_x];

    if (rad_val >= rad_threshold_ && !is_boosted_) {
      RCLCPP_WARN(this->get_logger(), "RADIATION (%d%%)! BOOSTING TO %.0f%% SPEED!", rad_val, boost_speed_pct_);
      is_boosted_ = true;
      publish_speed_limit(boost_speed_pct_);
    } else if (rad_val < rad_threshold_ && is_boosted_) {
      RCLCPP_INFO(this->get_logger(), "Radiation cleared. Restoring limit to %.0f%%.", normal_speed_pct_);
      is_boosted_ = false;
      publish_speed_limit(normal_speed_pct_);
    }
  }

  double rad_threshold_, normal_speed_pct_, boost_speed_pct_;
  bool is_boosted_, init_done_;
  nav_msgs::msg::OccupancyGrid::SharedPtr latest_map_;
  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Publisher<nav2_msgs::msg::SpeedLimit>::SharedPtr limit_pub_;
  std::unique_ptr<tf2_ros::Buffer> tf_buffer_;
  std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
  rclcpp::TimerBase::SharedPtr timer_check_;
  rclcpp::TimerBase::SharedPtr timer_init_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<AlaraSpeedReflex>());
  rclcpp::shutdown();
  return 0;
}